File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include<iostream>
#include<algorithm>
using namespace std;
int n, m;
int arr[16];
int cnt = 0;
void dfs(int val);
bool check(int node);

int main() {
    cin >> n;
 
    dfs(0);
    cout << cnt;
    return 0;
}

void dfs(int cur) {

    if (cur == n)
    {
        cnt++;
        return;
    }
    for (int i = 0; i < n; i++)
    {
        arr[cur] = i;
        if (check(cur)==true)
        {
            dfs(cur + 1);
        }
    }

}

bool check(int node)
{
    for (int i = 0; i < node; i++)
    {
        if (arr[node] == arr[i]||abs(arr[node]-arr[i])==node-i)
        {
            return false;
        }
    }
   
    return true;
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
from copy import deepcopy
import sys
import collections
arr=[0]*14
def check(i):
    for j in range(i):
        if arr[j]==arr[i] or abs(arr[j]-arr[i]) == abs(j-i):
            return False
    return True

def dfs(i):
    global ans
    if i==n:
        ans+=1
        return

    for j in range(n):
        arr[i]=j
        if check(i)==True:
            dfs(i+1)
        arr[i]=0


input=sys.stdin.readline
n=int(input())
ans=0
dfs(0)
print(ans)